<a href="https://colab.research.google.com/github/ccarpenterg/LearningTensorFlow2.0/blob/master/Gettingstarted_TensorFlow2_0_Intro_CNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Introduction to Convolutional Neural Networks with TensorFlow

In [0]:
!pip install tensorflow==2.0.0-alpha0

In [0]:
#import print function from future
from __future__ import absolute_import, division, print_function, unicode_literals

#import TensorFlow and check version
import tensorflow as tf

print(tf.__version__)

2.0.0-alpha0


**MNIST Fashion Dataset**

Downloading and feeding the dataset to the neural network is too simple. The dataset is conveniently included in the ketas module:

In [0]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11493376/11490434 [==============================] - 0s 0us/step


The image values are normalized by dividing them by 255.0, and therefore these values are cast to float numbers between 0.0 and 1.0.

**Building the Model for the Neural Network**

So now we are going to train a small neural network. We'll use the MNIST dataset to train our network, and in this case we have a hidden layer of 128 neurons and an output layer of 10 neurons (MNIST: 1, 2, 3, 4, 5, 6, 7, 8, 9, 0 digits).

Also, we are using Keras Flatten layer as the input layer. The MNIST dataset contains 28x28 images, we use the Flatten layer to turn those matrices into 784 (28x28) arrays. Now each hidden layer's neuron is connected to all 784 input layer's neurons.

In [0]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10, activation='softmax')
])

**Dropout Layer**

One of the main issues faced by ML engineers is overfitting. According to Wikipedia, overfitting is "the production of an analysis that corresponds too closely or exactly to a particular set of data, and may therefore fail to fit additional data or predict future observations reliably".

The process by which we prevent overfitting is called regularization.There are several methods for regularization; Dropout is a regularization technique that randomly drops out neurons during the training process.

In this particular network, we added a Dropout layer between the hidden layer (128 neurons) and the output layer (10 neurons). It has a droput rate of 0.2, which means that each neuron in the hidden layer has a probability of 20% of being dropped out by the network.

**Training Configuration: Optimizer, Loss and Metrics**

Now we need to choose the optimizer, the loss function and the metrics we are going to use to train our neural network.

In [0]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

Then we train and evaluate our model:

In [0]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 5s 91us/sample - loss: 0.0646 - accuracy: 0.9792
Epoch 2/5
60000/60000 [==============================] - 5s 91us/sample - loss: 0.0587 - accuracy: 0.9809
Epoch 3/5
60000/60000 [==============================] - 6s 95us/sample - loss: 0.0536 - accuracy: 0.9827
Epoch 4/5
60000/60000 [==============================] - 6s 98us/sample - loss: 0.0475 - accuracy: 0.9843
Epoch 5/5
10000/10000 [==============================] - 0s 45us/sample - loss: 0.0688 - accuracy: 0.9797


[0.06875320029149298, 0.9797]

**Results**

That's around 98% accuracy on the MNIST dataset. That's really outstanding for a NN with 3 layers (input, hidden and output). The droput is definitely an effective regularization technique.